### PR TITLE
SNOW-1708577 Parquet V2 support for new table format

### PR DESCRIPTION
**Pull request by sfc-gh-alhuang**
This PR aims to ensure consistency with the Iceberg table scanner and registration on the server side. It includes the following changes:

1. Added support for Parquet V2 (delta encoding) when the schema's `STORAGE_SERIALIZATION_POLICY` is set to `OPTIMIZED.`
2. Enabled dictionary encoding for Iceberg mode.

### CODE REVIEW

File: src/main/java/org/apache/parquet/hadoop/BdecParquetWriter.java
Comment (sfc-gh-hmadan) [inline src/main/java/org/apache/parquet/hadoop/BdecParquetWriter.java:47]: nit: add the final ones above the non-final ones!

File: src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java:103]: Use the BytesUtils methods from the parquet library instead?

File: src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java:102]: this is risky in that (a) the behavior of paddedChunkData can change and break the assumption here of no-padding-at-the-tail, (b) no guaranteed way of avoiding a corrupt value from flowing out.

1. lets extract extended metadata size before padding. 
2. verify that the last four bytes are == ParquetFileWriter.MAGIC
3. Is there nothing on ParquetFileWriter that exposes the extendedMetadataSize?

File: src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java:95]: we should do the extendedMetadataSize extraction after line 107 and make it conditional on internalParameterProvider.setIcebergSpecificFieldsInEp

File: src/main/java/net/snowflake/ingest/utils/Constants.java
Comment (sfc-gh-hmadan) [reply src/main/java/net/snowflake/ingest/utils/Constants.java:77]: lets use the constant from inside the parquet library instead of from here. Checkout ParquetFileWriter.MAGIC / MAGIC_STR.

File: src/main/java/net/snowflake/ingest/utils/Constants.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/utils/Constants.java:74]: Remove this hardcoded version and read the actual version from one of the parquet writers / parquet config classes ?

File: src/main/java/net/snowflake/ingest/utils/ParameterProvider.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/utils/ParameterProvider.java:53]: leaving it here so we don't forget it : PARQUET_WRITER_VERSION will be dynamic depending on the openchannelresponse

Slso don't make ENABLE_PARQUET_DICTIONARY_ENCODING into a public member. We don't want customers to be exposed to it just yet.

File: src/main/java/net/snowflake/ingest/utils/ParameterProvider.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/utils/ParameterProvider.java:274]: nit: add inline comment for argument name

File: src/main/java/net/snowflake/ingest/utils/Utils.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/utils/Utils.java:441]: (dupe) remove as we'll use parquet library's BytesUtils

File: src/main/java/net/snowflake/ingest/streaming/internal/InternalParameterProvider.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/InternalParameterProvider.java:35]: no need for a new method here, can reuse setIcebergSpecificFieldsInEp

File: src/main/java/net/snowflake/ingest/streaming/internal/EpInfo.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/EpInfo.java:25]: lets add this boolean into the ctor signature instead, as its a property of the columnEps that are already saved to this object's state.

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:653]: this looks odd to me. We are inferring an outer-class characteristic from an inner-object's state.
Lets take in a boolean alongside setDefaultValues

File: src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java:109]: we are in !getIsICebergMode (see line 100 above), so can just hardcode false here.

Comment (sfc-gh-hmadan) [reply src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java:109]: also applies to line 124.

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java:60]: is there any concurrent access to this collection?

File: src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java:198]: lets add a clientBufferParameters.getEnableDistinctValues() instead of getIsIcebergMode.

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java
Comment (sfc-gh-hmadan) [reply src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:653]: infact you can read this.clientBufferParameters.enableDistinctValues and don't even need an argument here.

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java:65]: this ctor is only called by test code. lets add isEnableDistinctValue to the signature, and make all the test classes run with both true and false.

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java:100]: nice!

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java:165]: ouch this will kill performance - both cpu and memory. If I'm not wrong we are already doing heavy validation of input strings, I'm hoping we can reuse that single pass on the whole string.

Another idea is to keep a `Map<String, Set<Int>>` here for string columns, the map key can be a CRC32 hash and the map value will be a list of row ids that contain a value with that hash. Only in the case where the Set has size > 1 do we look at the actual strings, before serialization. This is non-trivial amount of changes all over though.

File: src/main/java/net/snowflake/ingest/utils/Constants.java
Comment (rockwotj) [inline src/main/java/net/snowflake/ingest/utils/Constants.java:105]: Does this mean that non-iceberg tables (which are snowflake managed tables AFAIK) only support parquet v1?

File: src/main/java/org/apache/parquet/hadoop/BdecParquetWriter.java
Comment (sfc-gh-alhuang) [reply src/main/java/org/apache/parquet/hadoop/BdecParquetWriter.java:46]: Moved to top

File: src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java]: Changed to `ByteUtils` instead of own func.

File: src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java]: 1. The `chunkLength` returned by `padChunk` is the length before padding, the offset should be correct.
2. Added check in `Utils.getExtendedMetadataSize`.
3. Afaik the parquet file writer only log this value ([ref](https://github.com/apache/parquet-java/blob/eed26b4ea96881cc543b4233cdddece5d1d608d7/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java#L1918))

File: src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java:98]: Moved to `!getEnableChunkEncrpytion() == false` scope and guarded with `setIcebergSpecificFieldsInEp()`

File: src/main/java/net/snowflake/ingest/utils/Constants.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/Constants.java]: Yes we only do unencrypted footer which is `PAR1`. Removed this constant and use `ParuquetFileWriter.MAGIC_STR`

File: src/main/java/net/snowflake/ingest/utils/Constants.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/Constants.java]: Use major version from `ParquetFileWirter.CURRENT_VERSION`. Keep minor version as I cannot find one in `ParquetFileWriter`

File: src/main/java/net/snowflake/ingest/utils/ParameterProvider.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/ParameterProvider.java]: Changed to dynamic parquet writer version. Removed `ENABLE_PARQUET_DICTIONARY_ENCODING` , always enable dict encoding in iceberg mode.

File: src/main/java/net/snowflake/ingest/utils/ParameterProvider.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/ParameterProvider.java:266]: Added.

File: src/main/java/net/snowflake/ingest/utils/Utils.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/Utils.java]: Resolved.

File: src/main/java/net/snowflake/ingest/streaming/internal/InternalParameterProvider.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/InternalParameterProvider.java]: Removed.

File: src/main/java/net/snowflake/ingest/streaming/internal/EpInfo.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/EpInfo.java]: Moved to ctor.

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java]: Moved ndv flag to parameter. Cannot use `this.clientBufferParameters` as this is a static method called by chained of static methods.

File: src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java]: Switch to `getEnableDistinctValues()`

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java:59]: The value parser parse every row in a chunk in order. There's no concurrent access within the same `RowBufferStats`

File: src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java]: Added.

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java]: Added to signature and create parameterized tests.

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java]: Discussed offline. Lets keep NDV disabled by default. When it's enabled, skip the string ndv computation. Added a [JIRA](https://snowflakecomputing.atlassian.net/browse/SNOW-1721193) for this.

File: src/main/java/net/snowflake/ingest/utils/Constants.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/Constants.java:105]: The server-side scanner for FDN tables supports Parquet V2. This PR is specific to the Iceberg table feature and does not alter the default behavior for streaming to FDN tables.

Comment (rockwotj) [reply src/main/java/net/snowflake/ingest/utils/Constants.java:105]: Thanks!

File: src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java:144]: I'd much rather depend on our own constant than a third-party library's constant, I thought I had left a comment on this but don't see it anywhere :(

Comment (sfc-gh-hmadan) [reply src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java:144]: ok to take in next PR too, just remove the import whenever you revert this.

File: src/main/java/net/snowflake/ingest/streaming/internal/ClientBufferParameters.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/ClientBufferParameters.java:129]: this *might* be dependent on storage serialization policy too, lets verify. no need to hold up PR.

File: src/main/java/org/apache/parquet/hadoop/BdecParquetWriter.java
Comment (sfc-gh-hmadan) [inline src/main/java/org/apache/parquet/hadoop/BdecParquetWriter.java:43]: nit: lets rename this file to SnowflakeParquetWriter instead of BdecParquetWriter?!

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java:99]: nit: if (iceberg) ?

**Review by sfc-gh-hmadan — APPROVED** [review]
LGTM, left some minor nits/etc that can be taken with the next PR you have in flight. Lets merge this in to unblock that PR and other PRs that need to go in.

**Pull request merged**
